cloudinit/config/schemas/schema-cloud-config-v1.json

```diff
@@ -1924,22 +1939,32 @@
       "properties": {
         "packages": {
           "type": "array",
-          "description": "A list of packages to install. Each entry in the list can be either a package name or a list with two entries, the first being the package name and the second being the specific package version to install.",
+          "description": "An array containing either a package specification, or an object consisting of a package manager key having a package specification value . A package specification can be either a package name or a list with two entries, the first being the package name and the second being the specific package version to install.",
           "items": {
             "oneOf": [
               {
-                "type": "array",
-                "items": {
-                  "type": "string"
-                },
-                "minItems": 2,
-                "maxItems": 2
+                "type": "object",
+                "properties": {
+                  "apt": {
+                    "type": "array",
+                    "items": {
+                      "$ref": "#/$defs/package_item_definition"
+                    }
+                  },
+                  "snap": {
+                    "type": "array",
+                    "items": {
+                      "$ref": "#/$defs/package_item_definition"
+                    }
+                  }
+                }
               },
               {
-                "type": "string"
+                "$ref": "#/$defs/package_item_definition"
               }
             ]
           },
+          "additionalProperties": false,
```

Comment:
@TheRealFalcon I see you placed this additionalProperties:false here this doesn't end up applying to the oneOf object:
With your existing schema we still allow the following:
```
cat > pkg.yaml <<EOF
#cloud-config
packages:
 - apta: ["sl"]
EOF
$ PYTHONPATH=. python3 -m cloudinit.cmd.main schema -c pkg.yaml
Valid cloud-config: pkg.yaml
```

Reply:
That's not what I'm seeing:
```
root@me:~/a# cat some-config 
#cloud-config
packages:
 - apta: ["sl"]
root@me:~/a# cloud-init schema --config-file some-config
Invalid cloud-config some-config
Error: Cloud config schema errors: packages.0: Additional properties are not allowed ('apta' was unexpected), packages.0: {'apta': ['sl']} is not valid under any of the given schemas

Error: Invalid cloud-config schema: user-data
```
AFAIK, `additionalProperties` isn't valid inside the `properties` definition. It goes outside.